### PR TITLE
Add more complex slashing test code with random NCG and delegators.

### PR DESCRIPTION
**Pull request by s2quake**
This PR includes the following.

* Fixed issue where exception is thrown if slash factor is 0.
* Added slashing test code for random NCGs and random delegators.

The reason why I created this PR is to make sure that slashing is working correctly in a variety of situations (such as random NCGs and delegates).

### CODE REVIEW

File: .Lib9c.Tests/Action/DPoS/Control/SlashCtrlComplexTest.cs [hunk 1]
```diff
@@ -0,0 +1,588 @@
...
+        }
+
+        [Theory]
+        [MemberData(nameof(TestData))]
+        public void Slash_Test(bool jailed, int slashFactor)
+        {
+            // Given
+            var validatorNCG = NextNCG();
+            var validatorPower = PowerFromNCG(validatorNCG);
+            var validatorShare = ShareFromPower(validatorPower);
```
Comment (OnedgeLee) [inline .Lib9c.Tests/Action/DPoS/Control/SlashCtrlComplexTest.cs:55]: IMO, it's better to obtain `validatorPower` and `validatorShare` from states via `GetPower()` or `GetShare()` right after validator promotion, considering future modification on tests.
Actually, there are lots of lines like this, which seems to be fragile for me, considering test extension(for example, on the condition that inflation is added right after promotion) but only for this small unit test context, okay.

File: .Lib9c.Tests/Action/DPoS/Control/SlashCtrlTestBase.cs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        {
+            if (!value.Currency.Equals(Asset.ConsensusToken))
+            {
+                throw new ArgumentException(
+                        message: $"'{value}' is not {nameof(Asset.ConsensusToken)}",
+                        paramName: nameof(value));
+            }
+
+            return FungibleAssetValue.FromRawValue(Asset.Share, value.RawValue);
+        }
```
Comment (OnedgeLee) [inline .Lib9c.Tests/Action/DPoS/Control/SlashCtrlTestBase.cs:134]: `ShareFromPower` can make confusion, IMO.
`Power` is always equivalent to `GovernanceToken`, while `Share` isn't.
`Share` is conditionally equivalent to `GovernanceToken`, and existence of `ShareFromPower` method can make people to think those are exchangeable, though it's code for tests.

**Review by OnedgeLee — APPROVED** [review]
Left two comments, but I think those are trivial and matter of tastes for test, so approved.

**Review by limebell — APPROVED** [review]
LGTM

File: .Lib9c.Tests/Action/DPoS/Control/SlashCtrlComplexTest.cs (hunk 1 above)
Comment (s2quake) [reply .Lib9c.Tests/Action/DPoS/Control/SlashCtrlComplexTest.cs:55]: Except for `validatorNCG`, `validatorPower` and `validatorShare` are defined as expected values to be used in tests. Therefore, if a test passes with the expected and actual values being equal, I believe there is no fragility.

File: .Lib9c.Tests/Action/DPoS/Control/SlashCtrlTestBase.cs (hunk 2 above)
Comment (s2quake) [reply .Lib9c.Tests/Action/DPoS/Control/SlashCtrlTestBase.cs:134]: Hmm, is there any way to fix this? I have no idea right now, or is there a proper method name?

Comment (s2quake) [reply .Lib9c.Tests/Action/DPoS/Control/SlashCtrlTestBase.cs:134]: First, I think you're right.

**Pull request merged**